File: SocketClient/app/src/main/java/com/daeseong/socketclient/SocketManager.kt
package com.daeseong.socketclient

import android.util.Log
import kotlinx.coroutines.*
import org.json.JSONObject
import java.io.*
import java.net.InetSocketAddress
import java.net.Socket
import java.net.SocketException
import java.nio.ByteBuffer
import java.util.concurrent.atomic.AtomicBoolean

class SocketManager private constructor() : AutoCloseable {

    private val tag = SocketManager::class.java.simpleName

    private var socket: Socket? = null
    private val isConnected = AtomicBoolean(false)
    private var onReceiveListener: ((JSONObject) -> Unit)? = null
    private var onConnectListener: (() -> Unit)? = null
    private var onDisconnectListener: (() -> Unit)? = null
    private var onSendCompleteListener: ((Int) -> Unit)? = null
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    private var networkQuality: Float = 1.0f

    companion object {
        @Volatile
        private var instance: SocketManager? = null

        fun getInstance(): SocketManager =
            instance ?: synchronized(this) {
                instance ?: SocketManager().also { instance = it }
            }

        private const val RECONNECT_DELAY_MS = 5000L
        private const val MAX_RECONNECT_ATTEMPTS = 5
        private const val HEARTBEAT_INTERVAL_MS = 10000L
        const val CONNECTION_TIMEOUT_MS = 5000 // 5초
        private const val MAX_MESSAGE_SIZE = 100 * 1024 * 1024 // 100MB를 바이트 단위로 표현
    }

    private var serverIP: String = ""
    private var port: Int = 0
    private var reconnectAttempts = 0
    private var heartbeatJob: Job? = null

    suspend fun connect(serverIP: String, port: Int) {
        if (isConnected.get()) {
            throw IllegalStateException("이미 접속중")
        }

        this@SocketManager.serverIP = serverIP
        this@SocketManager.port = port
        reconnectAttempts = 0
        connectWithRetry()
    }

    private suspend fun connectWithRetry() {
        while (reconnectAttempts < MAX_RECONNECT_ATTEMPTS) {
            try {
                withContext(Dispatchers.IO) {
                    socket = Socket()
                    socket?.connect(InetSocketAddress(serverIP, port), CONNECTION_TIMEOUT_MS)
                    isConnected.set(true)
                }
                onConnectListener?.invoke()
                startListening()
                startHeartbeat()
                reconnectAttempts = 0
                return
            } catch (e: Exception) {
                Log.e(tag, "접속 실패: ${e.message}", e)
                reconnectAttempts++
                delay(RECONNECT_DELAY_MS)
            }
        }
        throw SocketException("최대 재접속 시도 횟수 ($MAX_RECONNECT_ATTEMPTS) 초과하여 접속 실패")
    }

    private fun startListening() {
        scope.launch {
            try {
                while (isConnected.get()) {
                    val message = withContext(Dispatchers.IO) {
                        readMessage()
                    }

                    if (message != null) {
                        withContext(Dispatchers.Main) {
                            onReceiveListener?.invoke(message)
                        }
                    } else {
                        Log.e(tag, "메시지 null")
                        break
                    }
                }
            } catch (e: Exception) {
                Log.e(tag, "수신 중 오류 발생: ${e.message}", e)
            } finally {
                reconnect()
            }
        }
    }

    private fun startHeartbeat() {
        heartbeatJob = scope.launch {
            while (isConnected.get()) {
                delay(HEARTBEAT_INTERVAL_MS)
                try {
                    val heartbeatMessage = JSONObject().apply {
                        put("type", "heartbeat")
                    }
                    send(heartbeatMessage)
                } catch (e: Exception) {
                    Log.e(tag, "하트비트 전송 실패: ${e.message}", e)
                    reconnect()
                }
            }
        }
    }

    private suspend fun readMessage(): JSONObject? {
        return withContext(Dispatchers.IO) {
            try {
                val lengthBuffer = ByteArray(4)
                val bytesRead = socket?.getInputStream()?.read(lengthBuffer)
                if (bytesRead != 4) return@withContext null

                val messageLength = ByteBuffer.wrap(lengthBuffer).int

                if (messageLength > MAX_MESSAGE_SIZE) {
                    Log.e(tag, "메시지 사이즈 초과")
                    return@withContext null
                }

                val messageBuffer = ByteArray(messageLength)
                var totalBytesRead = 0
                while (totalBytesRead < messageLength) {
                    val read = socket?.getInputStream()?.read(messageBuffer, totalBytesRead, messageLength - totalBytesRead)
                    if (read == -1) break
                    totalBytesRead += read ?: 0
                }
                if (totalBytesRead != messageLength) return@withContext null

                JSONObject(String(messageBuffer))
            } catch (e: SocketException) {
                Log.e(tag, "소켓 오류: ${e.message}", e)
                throw e // SocketException은 재연결을 시도해야 할 예외로 두고 상위 전달
            } catch (e: Exception) {
                Log.e(tag, "메시지 읽기 실패: ${e.message}", e)
                null
            }
        }
    }

    suspend fun send(message: JSONObject) {
        if (!isConnected.get()) {
            throw IllegalStateException("연결되지 않음")
        }

        try {
            val bytesSent = withContext(Dispatchers.IO) {
                val messageBytes = message.toString().toByteArray()
                val lengthPrefix = ByteBuffer.allocate(4).putInt(messageBytes.size).array()
                val outputStream = socket?.getOutputStream()
                outputStream?.write(lengthPrefix)
                outputStream?.write(messageBytes)
                outputStream?.flush()
                4 + messageBytes.size
            }

            withContext(Dispatchers.Main) {
                onSendCompleteListener?.invoke(bytesSent)
            }
        } catch (e: Exception) {
            Log.e(tag, "메시지 전송 실패: ${e.message}", e)
            reconnect()
            throw IOException("메시지 전송 실패: ${e.message}")
        }
    }

    fun setNetworkQuality(quality: Float) {
        networkQuality = quality
    }

    private fun disconnect() {
        if (!isConnected.getAndSet(false)) return

        try {
            socket?.close()
        } catch (e: Exception) {
            Log.e(tag, "연결 종료 중 오류 발생: ${e.message}", e)
        } finally {
            heartbeatJob?.cancel()
            scope.launch(Dispatchers.Main) {
                onDisconnectListener?.invoke()
            }
        }
    }

    fun setOnReceiveListener(listener: (JSONObject) -> Unit) {
        onReceiveListener = listener
    }

    fun setOnConnectListener(listener: () -> Unit) {
        onConnectListener = listener
    }

    fun setOnDisconnectListener(listener: () -> Unit) {
        onDisconnectListener = listener
    }

    fun setOnSendCompleteListener(listener: (Int) -> Unit) {
        onSendCompleteListener = listener
    }

    fun isConnected(): Boolean = isConnected.get()

    override fun close() {
        disconnect()
        scope.cancel()
    }

    private fun reconnect() {
        disconnect()
        scope.launch {
            try {
                connectWithRetry()
            } catch (e: Exception) {
                Log.e(tag, "재접속 실패: ${e.message}", e)
            }
        }
    }
}

File: dotnetMobileClient/SocketManager.cs
using System;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading.Tasks;

namespace dotnetMobileClient
{
    public class SocketManager : IDisposable
    {
        // 서버에 연결할 때의 타임아웃 시간(ms)
        private const int ConnectionTimeoutMs = 5000;

        // TCP 클라이언트 객체
        private TcpClient client;

        // 네트워크 스트림 객체
        private NetworkStream stream;

        // 연결 상태
        private bool isConnected = false;

        // 연결 완료
        private Func<Task> onConnectListener;

        // 연결 종료
        private Func<Task> onDisconnectListener;

        // 메시지 수신
        private Func<string, Task> onReceiveListener;

        // 메시지 전송 완료
        private Func<int, Task> onSendCompleteListener;


        // 서버 연결
        public async Task Connect(string serverIp, int port)
        {
            try
            {
                client = new TcpClient();
                var connectTask = client.ConnectAsync(serverIp, port);

                // 타임아웃 내에 연결을 시도
                if (await Task.WhenAny(connectTask, Task.Delay(ConnectionTimeoutMs)) == connectTask)
                {
                    await connectTask;
                    isConnected = true;
                    stream = client.GetStream();
                    if (onConnectListener != null) await onConnectListener();

                    Console.WriteLine($"서버에 연결됨: {serverIp}:{port}");
                }
                else
                {
                    throw new TimeoutException("연결 타임아웃");
                }
            }
            catch (Exception e)
            {
                Console.WriteLine($"연결 오류: {e.Message}");
                isConnected = false;
            }
        }

        // 연결 종료
        public async Task Disconnect()
        {
            if (stream != null) await stream.DisposeAsync();
            if (client != null) client.Close();
            isConnected = false;
            if (onDisconnectListener != null) await onDisconnectListener();

            Console.WriteLine("서버와의 연결이 끊어졌습니다.");
        }

        // 메시지 수신
        public async Task Listen()
        {
            var lengthBuffer = new byte[4]; // 메시지 길이를 저장할 버퍼
            while (isConnected)
            {
                try
                {
                    Console.WriteLine("메시지 수신 대기 중...");

                    int bytesRead = await stream.ReadAsync(lengthBuffer, 0, lengthBuffer.Length);
                    if (bytesRead == 0)
                    {
                        Console.WriteLine("연결이 종료되었습니다.");
                        break;
                    }

                    int length = IPAddress.NetworkToHostOrder(BitConverter.ToInt32(lengthBuffer, 0));

                    // 실제 메시지를 저장할 버퍼
                    var buffer = new byte[length]; 
                    bytesRead = await stream.ReadAsync(buffer, 0, buffer.Length);
                    string message = Encoding.UTF8.GetString(buffer, 0, bytesRead);

                    if (onReceiveListener != null) await onReceiveListener(message);
                }
                catch (Exception e)
                {
                    Console.WriteLine($"수신 오류: {e.Message}");
                    break;
                }
            }
            await Disconnect();
        }

        // 메시지 전송
        public async Task SendMessage(string message)
        {
            if (!isConnected) throw new InvalidOperationException("서버에 연결되어 있지 않음");

            var messageBytes = Encoding.UTF8.GetBytes(message);
            var lengthPrefix = BitConverter.GetBytes(IPAddress.HostToNetworkOrder(messageBytes.Length));
            await stream.WriteAsync(lengthPrefix, 0, lengthPrefix.Length); // 메시지 길이 전송
            await stream.WriteAsync(messageBytes, 0, messageBytes.Length); // 실제 메시지 전송
            
            if (onSendCompleteListener != null) await onSendCompleteListener(lengthPrefix.Length + messageBytes.Length);
            
            Console.WriteLine($"메시지 전송 완료: {message}");
        }

        
        public void SetOnReceiveListener(Func<string, Task> listener) => onReceiveListener = listener;
        public void SetOnConnectListener(Func<Task> listener) => onConnectListener = listener;
        public void SetOnDisconnectListener(Func<Task> listener) => onDisconnectListener = listener;
        public void SetOnSendCompleteListener(Func<int, Task> listener) => onSendCompleteListener = listener;

        // 객체 해제 시 연결 종료
        public void Dispose()
        {
            Disconnect().GetAwaiter().GetResult();
        }
    }
}
